py3dtilers/TilesetReader/parsed_b3dm.py

```diff
@@ -0,0 +1,130 @@
+        super().__init__(id)
+
+        self.geom = triangle_soup
+        self.set_box()
+
+    def set_material(self, mat_index, materials, tileset_path=None):
+        """
+        Set the material of this geometry.
+        :param mat_index: the index of the material
+        :param materials: the list of all the materials
+        :param tileset_path: the path to the tileset containing the texture image
+        """
+        self.material_index = mat_index
+        if materials[mat_index].is_textured():
+            path = os.path.join(tileset_path, "tiles", materials[mat_index].textureUri)
+            texture = Texture(path, self.geom.triangles[1])
+            self.set_texture(texture.get_texture_image())
+
+    def set_batchtable_data(self, bt_attributes):
+        """
+        Set the batch table data of this geometry.
+        :param bt_attributes: the batch table attributes for the whole tile
+        """
+        data = {}
+        index = int(self.id)
+        for attribute in bt_attributes:
+            if attribute == 'ids':
+                self.set_id(bt_attributes[attribute][index])
+            else:
+                data[attribute] = bt_attributes[attribute][index]
+        if data:
+            super().set_batchtable_data(data)
+
+
+class ParsedB3dms(ObjectsToTile):
+
+    def __init__(self, objects=None, tileset_paths_dict=None):
+        super().__init__(objects)
+        self.materials = []
+        self.tileset_paths_dict = tileset_paths_dict
+
+    def parse_materials(self, gltf):
+        """
+        Parse the materials from the glTF and create GlTFMaterials.
+        :param gltf: the gltf of the tile
+
+        :return: a list of GlTFMaterials
+        """
+        materials = gltf.header['materials']
+        gltf_materials = list()
+        self.mat_offset = len(self.materials)
+        for material in materials:
+            rgba = material['pbrMetallicRoughness']['baseColorFactor']
+            metallic_factor = material['pbrMetallicRoughness']['metallicFactor']
+            roughness_factor = material['pbrMetallicRoughness']['roughnessFactor']
+            if 'baseColorTexture' in material['pbrMetallicRoughness']:
+                index = material['pbrMetallicRoughness']['baseColorTexture']['index']
+                uri = gltf.header['images'][gltf.header['textures'][index]['source']]['uri']
+            else:
+                uri = None
+            gltf_materials.append(GlTFMaterial(metallic_factor, roughness_factor, rgba, textureUri=uri))
+        return gltf_materials
+
+    def parse_triangle_soup(self, triangle_soup, materials, tile_index=0):
+        """
+        Parse the triangle soup to re-create the geometries.
+        :param triangle_soup: the triangle soup
+        :param materials: the materials of the tile
+        :param tile_index: the index of the tile
+
+        :return: a list of geometries
+        """
+        triangles = triangle_soup.triangles[0]
```

Comment:
Man, the `TriangleSoup` class interface is really in need of some love.
Maybe we could suggest to Py3dTiles to e.g rename `.triangles[1]` accessor to a `get_vertex_ids()` method ?

Reply:
I completely agree. `.triangles` is a list containing the triangles, UVs, ids, and whatever the user puts in it. We should more getters/attributes in the `TriangleSoup` class in py3dtiles.